File: a_star/c/a_star.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#define INF INT_MAX

typedef struct Edge {
    int to_node;
    int weight;
    struct Edge* next;
} Edge;

typedef struct {
    int num_nodes;   
    Edge** adj_lists; 
} Graph;

typedef struct {
    int node;
    int f_score;
} HeapNode;

typedef struct {
    HeapNode* data;
    int size;
    int capacity;
} MinHeap;


MinHeap* create_heap(int capacity) {
    MinHeap* heap = (MinHeap*)malloc(sizeof(MinHeap));
    heap->data = (HeapNode*)malloc(capacity * sizeof(HeapNode));
    heap->size = 0;
    heap->capacity = capacity;
    return heap;
}

void swap(HeapNode* a, HeapNode* b) {
    HeapNode temp = *a;
    *a = *b;
    *b = temp;
}

void heap_push(MinHeap* heap, int node, int f_score) {
    if (heap->size == heap->capacity) return;

    int i = heap->size++;
    heap->data[i].node = node;
    heap->data[i].f_score = f_score;

    while (i != 0 && heap->data[(i - 1) / 2].f_score > heap->data[i].f_score) {
        swap(&heap->data[i], &heap->data[(i - 1) / 2]);
        i = (i - 1) / 2;
    }
}

HeapNode heap_pop(MinHeap* heap) {
    if (heap->size <= 0) {
        HeapNode empty = {-1, -1};
        return empty;
    }
    if (heap->size == 1) {
        heap->size--;
        return heap->data[0];
    }

    HeapNode root = heap->data[0];
    heap->data[0] = heap->data[heap->size - 1];
    heap->size--;

    int i = 0;
    while (1) {
        int smallest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < heap->size && heap->data[left].f_score < heap->data[smallest].f_score)
            smallest = left;
        if (right < heap->size && heap->data[right].f_score < heap->data[smallest].f_score)
            smallest = right;

        if (smallest == i) break;

        swap(&heap->data[i], &heap->data[smallest]);
        i = smallest;
    }

    return root;
}


Graph* create_graph(int num_nodes) {
    Graph* graph = (Graph*)malloc(sizeof(Graph));
    graph->num_nodes = num_nodes;
    graph->adj_lists = (Edge**)calloc(num_nodes, sizeof(Edge*));
    return graph;
}

void add_edge(Graph* graph, int src, int dest, int weight) {
    Edge* new_edge = (Edge*)malloc(sizeof(Edge));
    new_edge->to_node = dest;
    new_edge->weight = weight;
    
    new_edge->next = graph->adj_lists[src];
    graph->adj_lists[src] = new_edge;
}


int heuristic(int node, int goal) {
    return 0;
}

void a_star(Graph* graph, int start, int goal) {
    int* g_score = (int*)malloc(graph->num_nodes * sizeof(int));
    int* came_from = (int*)malloc(graph->num_nodes * sizeof(int));
    
    // Initialize
    for (int i = 0; i < graph->num_nodes; i++) {
        g_score[i] = INF;
        came_from[i] = -1;
    }

    g_score[start] = 0;

    MinHeap* open_set = create_heap(graph->num_nodes); 
    heap_push(open_set, start, 0);

    int found = 0;

    while (open_set->size > 0) {
        HeapNode currentObj = heap_pop(open_set);
        int current = currentObj.node;

        if (current == goal) {
            found = 1;
            break;
        }

        if (currentObj.f_score > g_score[current] + heuristic(current, goal)) {
            continue;
        }

        // Iterate neighbors
        Edge* adj = graph->adj_lists[current];
        while (adj != NULL) {
            int neighbor = adj->to_node;
            int weight = adj->weight;

            if (g_score[current] != INF && g_score[current] + weight < g_score[neighbor]) {
                g_score[neighbor] = g_score[current] + weight;
                came_from[neighbor] = current;
                
                int f_score = g_score[neighbor] + heuristic(neighbor, goal);
                heap_push(open_set, neighbor, f_score);
            }
            adj = adj->next;
        }
    }

    if (found) {
        printf("Path found: ");
        int* path_stack = (int*)malloc(graph->num_nodes * sizeof(int));
        int path_len = 0;
        
        int curr = goal;
        while (curr != -1) {
            path_stack[path_len++] = curr;
            curr = came_from[curr];
        }

        for (int i = path_len - 1; i >= 0; i--) {
            printf("%d ", path_stack[i]);
        }
        printf("\n");
        free(path_stack);
    } else {
        printf("No path exists between %d and %d.\n", start, goal);
    }

    free(g_score);
    free(came_from);
    free(open_set->data);
    free(open_set);
}


Graph* load_graph(const char* filename) {
    FILE* file = fopen(filename, "r");
    if (!file) {
        printf("Error: Could not open file %s\n", filename);
        return NULL;
    }

    char line[256];
    int max_node_id = 0;

    while (fgets(line, sizeof(line), file)) {
        if (line[0] == '#' || line[0] == '\n') continue;
        
        int u, v, w;
        if (sscanf(line, "%d %d %d", &u, &v, &w) == 3) {
            if (u > max_node_id) max_node_id = u;
            if (v > max_node_id) max_node_id = v;
        }
    }

    rewind(file);

    Graph* graph = create_graph(max_node_id + 1);

    // PASS 2: Load data
    while (fgets(line, sizeof(line), file)) {
        if (line[0] == '#' || line[0] == '\n') continue;
        
        int u, v, w;
        if (sscanf(line, "%d %d %d", &u, &v, &w) == 3) {
            add_edge(graph, u, v, w);
        }
    }

    fclose(file);
    return graph;
}


int main() {
    printf("Loading graph...\n");
    Graph* my_graph = load_graph("D:/CODE/OESK/graph_with_weights.txt");

    if (my_graph) {
        printf("Graph loaded. Max Node ID: %d\n", my_graph->num_nodes - 1);

        int start_node = 0;
        int goal_node = 425875;

        if (start_node < my_graph->num_nodes && goal_node < my_graph->num_nodes) {
            printf("Searching path from %d to %d...\n", start_node, goal_node);
            a_star(my_graph, start_node, goal_node);
        } else {
            printf("Start or Goal node is out of bounds.\n");
        }

        for(int i=0; i < my_graph->num_nodes; i++) {
            Edge* curr = my_graph->adj_lists[i];
            while(curr) {
                Edge* temp = curr;
                curr = curr->next;
                free(temp);
            }
        }
        free(my_graph->adj_lists);
        free(my_graph);
    }

    return 0;
}

File: a_star/a_star.js
const fs = require('fs');
const readline = require('readline');

class MinHeap {
    constructor() {
        this.heap = [];
    }

    push(val) {
        this.heap.push(val);
        this.bubbleUp(this.heap.length - 1);
    }

    pop() {
        if (this.heap.length === 0) return null;
        const min = this.heap[0];
        const end = this.heap.pop();
        if (this.heap.length > 0) {
            this.heap[0] = end;
            this.bubbleDown(0);
        }
        return min;
    }

    size() {
        return this.heap.length;
    }

    bubbleUp(n) {
        while (n > 0) {
            let parent = Math.floor((n - 1) / 2);
            if (this.heap[n].f >= this.heap[parent].f) break;
            [this.heap[n], this.heap[parent]] = [this.heap[parent], this.heap[n]];
            n = parent;
        }
    }

    bubbleDown(n) {
        let length = this.heap.length;
        while (true) {
            let left = 2 * n + 1;
            let right = 2 * n + 2;
            let smallest = n;

            if (left < length && this.heap[left].f < this.heap[smallest].f) smallest = left;
            if (right < length && this.heap[right].f < this.heap[smallest].f) smallest = right;
            
            if (smallest === n) break;
            
            [this.heap[n], this.heap[smallest]] = [this.heap[smallest], this.heap[n]];
            n = smallest;
        }
    }
}


function heuristic(node, goal) {
    return 0; 
}

function reconstructPath(cameFrom, current) {
    const totalPath = [current];
    while (cameFrom[current] !== undefined) {
        current = cameFrom[current];
        totalPath.push(current);
    }
    return totalPath.reverse();
}

function aStar(graph, start, goal) {
    const openSet = new MinHeap();
    openSet.push({ f: 0, node: start });

    const cameFrom = {};
    
    const gScore = {}; 
    gScore[start] = 0;

    
    while (openSet.size() > 0) {
        const { f: currentF, node: current } = openSet.pop();

        if (current === goal) {
            return reconstructPath(cameFrom, current);
        }

        const currentG = gScore[current] !== undefined ? gScore[current] : Infinity;
        if (currentF > currentG + heuristic(current, goal)) {
            continue;
        }

        const neighbors = graph[current];
        if (!neighbors) continue;

        for (const neighborStr in neighbors) {
            const neighbor = parseInt(neighborStr);
            const weight = neighbors[neighbor];
            
            const tentativeGScore = currentG + weight;
            const neighborG = gScore[neighbor] !== undefined ? gScore[neighbor] : Infinity;

            if (tentativeGScore < neighborG) {
                cameFrom[neighbor] = current;
                gScore[neighbor] = tentativeGScore;
                
                const fScore = tentativeGScore + heuristic(neighbor, goal);
                openSet.push({ f: fScore, node: neighbor });
            }
        }
    }

    return null;
}

async function loadGraph() {
    const graph = {};

    const fileStream = fs.createReadStream("D:/CODE/OESK/graph_with_weights.txt");

    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    console.log("Loading graph...");

    for await (const line of rl) {
        const trimmedLine = line.trim();
        
        if (trimmedLine.startsWith("#") || !trimmedLine) {
            continue;
        }

        // Split by whitespace
        const parts = trimmedLine.split(/\s+/);

        if (parts.length !== 3) {
            continue;
        }

        const fromNode = Number(parts[0]);
        const toNode = Number(parts[1]);
        const weight = Number(parts[2]);

        // Initialize objects if they don't exist
        if (!graph[fromNode]) graph[fromNode] = {};
        if (!graph[toNode]) graph[toNode] = {}; 

        graph[fromNode][toNode] = weight;
    }

    console.log("Graph loaded.");
    return graph;
}

(async () => {
    const myGraph = await loadGraph();
    const nodeCount = Object.keys(myGraph).length;

    if (nodeCount > 0) {
        console.log(`Graph loaded with ${nodeCount} nodes.`);

        const startNode = 0;
        const goalNode = 425876;

        // Check existence
        if (myGraph[startNode] && myGraph[goalNode]) {
            const path = aStar(myGraph, startNode, goalNode);
            if (path) {
                console.log("Path found:", path);
            } else {
                console.log("No path exists.");
            }
        } else {
            console.log("Start or Goal node not in graph.");
        }
    }
})();